### PR TITLE
Updates to colored group indicator for cited entries

### LINKED ISSUE #6394: Main table: Coloured group indicator for cited entries does not show up consistently anymore
Tested with the latest JabRef development version (master branch) on Ubuntu 16.04

Unfortunately, the colored group indicator for cited entries does not show up consistently anymore in the main table: The badge icon with the number of cited entries works, but the colored group indicator in the main table does not show up for most of the cited entries. Sometimes only one, two or four indicators show up. (Updating the group repeatedly or restarting JabRef does not help either.)

@tobiasdiez I strongly suspect that the cause of this issue is connected with the recently conducted performance optimizations of the group functionality (faster group calculations). When reverting back for two weeks, this problem vanishes.

1. Create group of cited entries by linking the corresponding `.aux` file
1. Cite entries in a tex file and build the pdf file
2. Update group of cited entries
3. Observation: Not all indicators show up

### PR DESCRIPTION
**Pull request by k3KAW8Pnf7mkmdSMPHz27**
Fixes #6394.

An entry can belong to a specific group depending on,
1. Field or type in the `BibEntry`
2. The settings of the specific group

Hence, ~any changes in a field or type in the `BibEntry`, or~ any change in the groups should reevaluation wether or not the entry is still in any given group.

I am not very comfortable with the JavaFX/MVVC structure yet, so I have attempted an approach with as few structural changes as possible.

- [x] Keep the count "badge" is up-to-date
- [x] Refresh the selected groups
  - [x] Deal with `ConcurrentModificationException`
  - [x] ~Fix performance bottleneck~
    - [x] ~Occurs when switching library and the group is selected~
- [x] Check for potential memory leaks,
  - [x] caused by extending `Observable` in `TexGroup.java` (how should the listeners be stored?)
  - [x] dependence on deprecated method `GroupTreeNode#setGroup(AbstractGroup newGroup)`
  - [x] use of `EventBus` without unregistering listener
    - [x] Make a WeakInvalidationListener adapter?
- [x] JabRef believes the .bib file has changed when the .aux file changes
  - [x] Check if the `isFilteredOut` attribute of `BibDatabaseContextChangedEvent` is used to prevent saving
    - [x] Add JavaDoc
- [x] ~How does the current `GroupInvalidationEvent` affect the hierarchical group structure? (should it be fired from elsewhere so that only partial recalculations of group colors are necessary)~
- [x] Replace unnecessary adapter class

---

- [x] Change in CHANGELOG.md described (if applicable)
- [x] ~Tests created for changes (if applicable)~
- [x] Manually tested changed features in running JabRef (always required)
- [x] ~Screenshots added in PR description (for UI changes)~
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Would any of the following be more/less valid approaches? 

1. Make `TexGroup` implement `Observable`,
2. Make `TexGroup.keysUsedInAux` an `ObservableSet` (I'd argue this seem similar to how `BibEntry#getObservables` operates)

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I added a minimum working example of implementing the `Observable` interface.  Any comments/suggestions are welcome (including scratch this approach X) )

**Comment by Siedlerchr** [comment]
Thanks for tackling this! It would be interesting to see how it performs with many group (can provide you a test file via mail or so, cause its confidential)

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@Siedlerchr thank you for reviewing my PRs and commits!

I unfortunately do not have a database with many groups. I am exclusively using the JabRef test databases (in this case `origin.bib` and `paper.aux`.

If the approach of implementing `Observable` is ok within the MVVM framework, I can look through the code in more detail to see what impact using `notifyAboutDescendantChange` is likely to have (e.g., will all changes be lazily evaluated).

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I can't observe impactful slowdowns/memory leaks, but there shouldn't be many changes that affects the large database. Most potential memory leaks will only occur when an `.aux` file is linked.

I have update the description with what I think is left, and commented on a change that can mostly be reverted if performance is an issue. Sorry about the slow going, JavaFX is somewhat new to me.

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 1]
```diff
@@ -68,18 +71,29 @@ public BibEntry getEntry() {
...
-                                                 .map(GroupTreeNode::getGroup)
-                                                 .collect(Collectors.toList());
-                groups.remove(root.get().getGroup());
-                return groups;
-            });
-        }
-        return new SimpleObjectProperty<>(Collections.emptyList());
+        ObservableValue<List<AbstractGroup>> groupsBinding = new ObjectBinding<List<AbstractGroup>>() {
+            {
+                bind(entry.getFieldsObservable(), entry.typeProperty());
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:76]: Here we can change the dependence to only be on StandardField.GROUPS again. It could be argued that this change is out of the scope of this PR and has potential performance effects.

**Comment by DominikVoigt** [comment]
Hey there!
Thank you for your contribution.
Seeing that your last commit was pushed 5 days ago,
I want to ask whether you are finished with your work?
Is this PR ready for review :)?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Hey @DominikVoigt . I am sorry but no, I thought I would make a 30 min fix in #7173 but unfortunately it didn't turn out that way o.O
I don't think this is worth reviewing yet, the `JabRef believes the .bib file has changed when the .aux file changes` bullet item most likely means I am going to have to change/extend the metadata updates which might/might not be ok from JabRef's point-of-view.

**Comment by Siedlerchr** [comment]
@k3KAW8Pnf7mkmdSMPHz27  The bib file is changed can be also #5967

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Thank you for the link!  
It might be both o.O
I should have time to continue on this later today *knock on wood*

**Comment by Siedlerchr** [comment]
As a first workaround try disabling autosave.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@Siedlerchr at least for now it happens independently of autosave.

The reason why JabRef believes that the database has changed is that I wanted to use `GroupUpdatedEvent` to notify any listener that a group has changed. `GroupUpdatedEvent` extends `BibDatabaseContextChangedEvent` which I guess triggers

https://github.com/JabRef/jabref/blob/08e824f506a997e9f3f2e87f70fa256d1ba40620/src/main/java/org/jabref/gui/LibraryTab.java#L356-L359

among others. (I wanted to view it as a change to the `MetaData`)
The issue with `TexGroup` is that it

1. requires an update without any interaction from the user (it updates on changes in an external file), and
2. these updates does not need to be saved.

I think it is the only group with these characteristics, hence I don't see it motivated to extend the `GroupUpdatedEvent` since it seems well established that `BibDatabaseContextChangedEvent` implies a change that requires saving.

I'll figure something out but I am temporarily a bit stuck 😱

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Hum... `TexGroup` has a reference to the `MetaData`. Is it ok if I its `EventBus` to post updates instead?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Barring changing quite a large part of the code to use Observables (I saw some discussion regarding this in other issues but I am not aware of if there is any consensus?), this is my attempt at addressing this issue. Sorry about the delay and rather large changes.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
D'oh, I just realised a mistake I'll have to fix tomorrow =/

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
It might make sense to make `TexGroup` observable or to use its `FileUpdateMonitor` directly in the `GroupNodeViewModel`.

File: src/main/java/org/jabref/model/groups/TexGroup.java [hunk 2]
```diff
@@ -119,6 +120,7 @@ public Path getFilePath() {
     public void fileUpdated() {
         // Reset previous parse result
         keysUsedInAux = null;
+        DefaultTaskExecutor.runInJavaFXThread(() -> metaData.groupsBinding().invalidate());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/groups/TexGroup.java:123]: In the best case, the invalidation would bubble up automatically from the group along the whole tree. So somehing along the following lines:
- Make each `Group` an observable value that is invalidated when it's properties change. Could be done by making AbstractGroup derive from `PreboundBinding`, change `keysUsedInAux` to an observable set, and invoke `super(keysUsedInAux)` in the constructor
- Invalidate the grouptreenode as soon as the group is invalidated. Should be possible by letting TreeNode derive from PreboundBinding, and then use an extractor here https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/model/TreeNode.java#L59 to invalidate the list, when one of the nodes changes.
- Then the root should be invalidated whenever one of the groups reports that is invalidate/changed.

However, that is more work, so I'm fine with this top-down-approach for now as well.

**Review by tobiasdiez — COMMENTED** [review]
Do you still have open questions or is this already ready for review?

File: src/main/java/org/jabref/model/groups/TexGroup.java (hunk 2 above)
Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/groups/TexGroup.java:123]: I think I'll skip making it observable. It feels like a rabbit-hole that I don't believe I have the JavaFX know-how to dig myself out of (yet). Making `GroupTreeNode#getMatchingGroups` observable after extending with `PreboundBinding` would be another low hanging fruit.

The remaining question would be if from a MVVC point-of-view, does it make more sense to implement the `fileUpdate` method https://github.com/JabRef/jabref/blob/afdb1942c5ec2126e69c92c7f2660459a7bdbeae/src/main/java/org/jabref/model/groups/TexGroup.java#L119
in the `GroupNodeViewModel`, i.e., "initialize" it similar to the `AutomaticPersonsGroup`
https://github.com/JabRef/jabref/blob/afdb1942c5ec2126e69c92c7f2660459a7bdbeae/src/main/java/org/jabref/gui/groups/GroupNodeViewModel.java#L73

Comment (tobiasdiez) [reply src/main/java/org/jabref/model/groups/TexGroup.java:123]: No, it's fine that the texgroup invalidates the root - that's actually good design, as both of these life in the model. You don't want to take care of this through the UI classes (ViewModel).

Can you please remove the `DefaultTaskExecutor.runInJavaFXThread` part. The BindingsHelper has a `forUI` helper method that allows you consume change events on the JavaFX thread.

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/groups/TexGroup.java:123]: Just so that I understand you.
The `BindingsHelper.forUI` accepts an `ObservableList` and wraps it in a `UiThreadList` -> I should create an `UiThreadOptionalBinding` wrapper with analogous use to `UiThreadList`/`UiThreadObservableList` and wrap `groupsRootBinding` in it?

I am guessing that means it is ok that any test case that use `setGroup` will get an `Exception in thread "Test worker" java.lang.IllegalStateException: Toolkit not initialized`? (at least that is what happens in `ConvertMarkingToGroupsTest#performMigrationForSingleEntry` on my initial attempt)
I mean, any change in the groups _should_ trigger an UI update.

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/groups/TexGroup.java:123]: Sure! I'll create an `UiOptionalBinding` wrapper and add a `BindingsHelper#forUI` method for it.

Comment (calixtus) [reply src/main/java/org/jabref/model/groups/TexGroup.java:123]: I remember we talked once in a devcall about abandoning the google event bus and using observables or properties instead, but I think this is stuff for some far future changes...

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/groups/TexGroup.java]: I might return and look more at it for groups at a later point. I just feel that my JavaFX knowledge is not up to task for it yet. I need to experiment with it more, on something less impactful 😝

**Comment by calixtus** [comment]
Hey @k3KAW8Pnf7mkmdSMPHz27 , you didn't change anything on this PR for 5 days? Is this PR ready for review?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Sorry. Yes, it is ready for review

File: src/main/java/org/jabref/model/metadata/MetaData.java [hunk 3]
```diff
@@ -9,7 +9,11 @@
 import java.util.Objects;
 import java.util.Optional;
 
+import javafx.beans.property.ObjectProperty;
+import javafx.beans.property.SimpleObjectProperty;
+
 import org.jabref.architecture.AllowedToUseLogic;
+import org.jabref.gui.util.uithreadaware.UiThreadOptionalBinding;
```
[the remark this answers is not in the document]
Comment (tobiasdiez) [reply src/main/java/org/jabref/model/metadata/MetaData.java:16]: @JabRef/developers Our architecture tests don't work anymore?! This is clearly a violation but the tests pass.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Thanks! 

The code looks very good to me. The only problem I see is that the UI-binding already happens in the model class. This should be moved to the gui classes that need such a wrapping.

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 4]
```diff
@@ -67,19 +69,15 @@ public BibEntry getEntry() {
...
-                                                 .stream()
-                                                 .map(GroupTreeNode::getGroup)
-                                                 .collect(Collectors.toList());
-                groups.remove(root.get().getGroup());
-                return groups;
-            });
-        }
-        return new SimpleObjectProperty<>(Collections.emptyList());
+    private static Binding<List<AbstractGroup>> createMatchedGroupsBinding(BibDatabaseContext database, BibEntry entry) {
+        return new UiThreadBinding<>(EasyBind.combine(entry.getFieldBinding(StandardField.GROUPS), database.getMetaData().groupsBinding(),
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:73]: Should this binding be modified to depend on any changes in the bibentry? #7325 depends on keyword field, instead of group field. Free text search depends on all fields. Should it be in a different PR? Personally I'd prefer if `createMatchedGroupsBinding` could be changed to return a `FilteredList`, but that would likely take me a while to implement...

**Review by k3KAW8Pnf7mkmdSMPHz27 — COMMENTED** [review]
I'd appreciate an opinion on what should create an update of the group color region in the maintable (`entry.getObservables()`?). Also, should that part of #7325 be addressed in this PR? (the group color region not updating)

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java (hunk 4 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:73]: Yes, I think it's a good idea to listen to changes of the entry. Let's discuss this in a follow-up PR, since this one looks already good.

**Review by tobiasdiez — APPROVED** [review]

File: src/main/java/org/jabref/model/metadata/MetaData.java [hunk 5]
```diff
@@ -51,7 +57,8 @@
     private final Map<EntryType, String> citeKeyPatterns = new HashMap<>(); // <BibType, Pattern>
     private final Map<String, String> userFileDirectory = new HashMap<>(); // <User, FilePath>
     private final Map<String, Path> laTexFileDirectory = new HashMap<>(); // <User, FilePath>
-    private GroupTreeNode groupsRoot;
+    private final ObjectProperty<GroupTreeNode> groupsRoot = new SimpleObjectProperty<>(null);
```
[the remark this answers is not in the document]
Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/metadata/MetaData.java:59]: I'll have to think more about this one. I keep ending up with `OptionalWrapper<Optional<GroupTreeNode>>` 😛

[the remark this answers is not in the document]
Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/metadata/MetaData.java:59]:
```
OptionalObjectProperty<GroupTreeNode> groupsRoot = OptionalObjectProperty.empty();
OptionalBinding<GroupTreeNode> groupsRootBinding = new OptionalWrapper<>(groupsRoot.orElse(null));
```
This is what ends up making the most sense to me. I'll add it to my todo-list for the follow up.

**Review by calixtus — APPROVED** [review]
Worked myself through the changes. Looks good also to me, so I will merge this PR.
Thanks for your work on this issue!

**Comment by systemoperator** [comment]
I have just tried this fix with the current master build on Ubuntu 16.04. Unfortunately, it seems, neither the colored group indicator, nor the badge icon are updated automatically in my situation:

- Step 1: Build tex file with additional reference. Observations:
    - **Counter in badge icon for cited entries is not incremented.**
    - **Badge icon for cited entries is not highlighted when the bib entry is selected.**
    - **Colored group indicator for cited entries is not shown for the newly added reference.**
- Step 2: Update group of cited entries: Just click edit the group of cited entries and save it again. When a dialogue appears choose, to assign the original group's entries again to this group. Observations:
    - Counter in badge icon for cited entries is now incremented properly.
    - Badge icon for cited entries is now highlighted properly when the bib entry is selected.
    - **Colored group indicator for cited entries is still not shown for the newly added reference.**
- Step 3: Close and reopen the bib-file. Observations:
    - Everything is shown as expected.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Sorry about that. I am using Texpad with Mac OS X and the group gets updated of the .aux file as I'd expect on JabRef 5.3--2021-01-26--034cf8c.
Could you give me more details on "Step 1" (which application you are using etc.) and I'll try to get a virtual box with Ubuntu up and going in the coming days?

**Comment by systemoperator** [comment]
I have used a test `.tex` file with biblatex (see first example in https://www.overleaf.com/learn/latex/bibliography_management_with_biblatex). The library was configured for biblatex in JabRef as well. I have used TeXstudio, but I think it's independent of this, because I have even simply opened the `.aux` file with some editor and I have directly removed or added entries like `\abx@aux@cite{XXX2008}`. It seems, that the file monitor does not get triggered properly on a file change, thus, `keysUsedInAux` will not be cleared and the current group bindings are not invalidated.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> directly removed or added entries like `\abx@aux@cite{XXX2008}`. It seems, that the file monitor does not get triggered properly on a file change, thus, `keysUsedInAux` will not be cleared and the current group bindings are not invalidated.

Yup. Adding/removing entries in that fashion works for me in OS X. I'll look into it. I don't really have a hypothesis for why it doesn't work in Ubuntu (yet)

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Is it possible your editor is generating [a "create event" instead of a "modify event"](https://stackoverflow.com/questions/43447654/ubuntu-java-watchservice-is-broken)?

You are using a local drive? (not remote)

**Comment by systemoperator** [comment]
Yes, on a local drive. You are right: I have created a small java test program of your provided link.

- Creating the file "testFile.txt" triggers:

`ENTRY_CREATE:testFile.txt`

- Editing the file with a simple text editor triggers:

`ENTRY_CREATE:testFile.txt`

- Editing the file with TeXstudio triggers:

```
ENTRY_CREATE:testFile.txt.G29838
ENTRY_MODIFY:testFile.txt.G29838
ENTRY_MODIFY:testFile.txt.G29838
ENTRY_DELETE:testFile.txt.G29838
ENTRY_CREATE:testFile.txt
```

**Comment by systemoperator** [comment]
@tobiasdiez @k3KAW8Pnf7mkmdSMPHz27 Do you think, that it would make sense to also observe the `ENTRY_CREATE` event to make the detection of file-modifications more reliable, since two out of three of my tested editors trigger the `ENTRY_CREATE` event and other programs apparently also check on this event?

**Comment by tobiasdiez** [comment]
Yes, I guess it's a good idea to also listen for entry_create. At least I cannot think of a negative sideeffect.
